### PR TITLE
update collections

### PR DESCRIPTION
**Pull request by mysqto**
Changes:
1. Change `Filter` and `All` from `comparable` to `any`
2. Add `map` related functions like `ContainsKeys` `ContainsValues`

### CODE REVIEW

**Review by viettd — APPROVED** [review]

**Review by redho-wego — DISMISSED** [review]

File: collection/collection.go [hunk 1]
```diff
@@ -97,3 +97,84 @@ func Equal[T comparable](s1 []T, s2 []T) bool {
...
 	return true
 }
+
+// ContainsKeys checks whether all items in the slice are in the map
+func ContainsKeys[M ~map[K]V, K comparable, V any](m M, ks []K) bool {
+	for _, v := range ks {
+		if _, ok := m[v]; !ok {
+			return false
+		}
+	}
```
Comment (viettd) [inline collection/collection.go:107]:
```suggestion
	for _, k := range ks {
		if _, ok := m[k]; !ok {
			return false
		}
	}
```

File: collection/collection.go [hunk 2]
```diff
@@ -97,3 +97,84 @@ func Equal[T comparable](s1 []T, s2 []T) bool {
...
+	return true
+}
+
+// ContainsAnyKeys checks whether any item in the slice is in the map
+func ContainsAnyKeys[M ~map[K]V, K comparable, V any](m M, ks []K) bool {
+	for _, v := range ks {
+		if _, ok := m[v]; ok {
+			return true
+		}
+	}
```
Comment (viettd) [inline collection/collection.go:128]:
```suggestion
	for _, k := range ks {
		if _, ok := m[k]; ok {
			return true
		}
	}
```

File: collection/collection.go [hunk 3]
```diff
@@ -97,3 +97,84 @@ func Equal[T comparable](s1 []T, s2 []T) bool {
...
+	return false
+}
+
+// ContainsNoneKeys checks whether none of the items in the slice is in the map
+func ContainsNoneKeys[M ~map[K]V, K comparable, V any](m M, ks []K) bool {
+	for _, v := range ks {
+		if _, ok := m[v]; ok {
+			return false
+		}
+	}
```
Comment (viettd) [inline collection/collection.go:149]:
```suggestion
	for _, k := range ks {
		if _, ok := m[k]; ok {
			return false
		}
	}
```

File: collection/collection.go [hunk 4]
```diff
@@ -97,3 +97,84 @@ func Equal[T comparable](s1 []T, s2 []T) bool {
...
+	return true
+}
+
+// Keys returns the keys of the map
+func Keys[M ~map[K]V, K comparable, V any](m M) []K {
+	vs := make([]K, 0)
+	for k := range m {
+		vs = append(vs, k)
+	}
+	return vs
```
Comment (viettd) [inline collection/collection.go:170]:
```suggestion
	ks := make([]K, 0)
	for k := range m {
		ks = append(ks, k)
	}
	return ks
```

File: collection/collection.go [hunk 5]
```diff
@@ -97,3 +97,84 @@ func Equal[T comparable](s1 []T, s2 []T) bool {
 
 	return true
 }
+
+// ContainsKeys checks whether all items in the slice are in the map
+func ContainsKeys[M ~map[K]V, K comparable, V any](m M, ks []K) bool {
```
Comment (viettd) [inline collection/collection.go:102]: Can we rename all `Containsxxx` to `Containxxx` including the current `Contains`?

Comment (viettd) [reply collection/collection.go:102]: >JAVA also have the same naming/
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/List.html#containsAll(java.util.Collection)

Oh, I never realize about this inconsistency: `containsAll`& `equals` have `s` but others (`addAll`, `removeAll`, `replaceAll`, ...) don't. Do you know why?

>I think it's OK to use Contains bc the subject array and map is a singular

Yeah, I think it's ok to use too. But if we use the same logic then shouldn't a lot of other functions/methods should be in the same form as well?

Go also has `s` for `Contains` for some reason: https://pkg.go.dev/golang.org/x/exp/slices#Contains

Comment (viettd) [reply collection/collection.go:102]: Ah, your explanation makes sense to me.
I'm ok to change `Equal` to `Equals` too as it fits your thinking above.
Seems like Go team missed it 😂 https://pkg.go.dev/golang.org/x/exp/slices#Equal

**Review by viettd — APPROVED** [review]

**Review by redho-wego — APPROVED** [review]

**Pull request merged**